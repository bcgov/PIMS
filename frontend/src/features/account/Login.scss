@import '../../styles.scss';
@import '../../colors.scss';
@import '../../variables.scss';

.login {
  height: calc(100vh - #{$footer-height} - #{$header-height});
  background-image: url('../../assets/images/SeasideBlue.png');
  background-size: cover;
  background-position: center top;

  .unauth {
    font-size: 24px;
    margin: 91px auto;
    background-color: rgba(250, 250, 250, 0.9); //allows for slight transparency
    padding-top: 15px;
    padding-bottom: 20px;
    max-width: 750px;
    border-radius: 4px;
    .sign-in {
      justify-content: center;
    }
    .block {
      background-color: $filter-background-color;
      padding: 20px 0;
      max-width: 650px;
      max-height: 475px;
      text-align: center;
      border-radius: 4px;
      min-width: 90%;
    }
    h1 {
      padding-bottom: 18px;
      font-family: 'BCSans', Fallback, sans-serif;
      font-weight: 700;
      font-size: 24px;
      font-style: italic;
      color: $primary-color;
    }
    h6 {
      margin-bottom: 20px;
    }
    p {
      font-size: 12px;
      color: $text-color;
      font-family: 'BCSans', Fallback, sans-serif;
      line-height: 130%;
      margin-left: 30px;
      margin-right: 30px;
      margin-bottom: 2px;
    }
    .btn {
      margin-top: 30px;
      margin-bottom: 30px;
    }
    .btn-link {
      margin-left: 30px;
      margin-top: 10px;
      font-size: 12px;
    }
    .btn-link:focus,
    .btn-link:active {
      background: none !important;
    }
    .bceid {
      margin-bottom: -30px;
    }
    .jumbotron {
      background-color: white;
      padding: 20px;
      margin-left: 40px;
      margin-right: 40px;
      width: 90%;
      justify-content: left;
      border: 1px;
      border-style: solid;
      border-color: #707070;
      text-align: left;
      p {
        margin: 10px;
      }
    }
    .blockText {
      margin: 0 auto 38px auto;
    }
    .or {
      margin: 20px auto;
    }
    .border-dark {
      border-color: $primary-color !important;
      color: $text-color !important;
    }
    .pims-logo {
      margin-left: -300px;
      margin-bottom: 10px;
    }
  }
}
