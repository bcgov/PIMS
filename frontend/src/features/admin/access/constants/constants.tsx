import { ColumnWithProps } from 'components/Table';
import { AccessStatusDisplayMapper } from 'constants/accessStatus';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IAccessRequestModel } from '../interfaces';

export const columnDefinitions: ColumnWithProps<IAccessRequestModel>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IAccessRequestModel>) => {
      return (
        <Link to={`/admin/user/${props.row.original.userId}`}>{props.row.original.username}</Link>
      );
    },
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Status',
    accessor: 'status',
    align: 'left',
    clickable: true,
    width: 100,
    Cell: (props: CellProps<IAccessRequestModel>) =>
      AccessStatusDisplayMapper[props.row.original.status],
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Role',
    accessor: 'role',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
