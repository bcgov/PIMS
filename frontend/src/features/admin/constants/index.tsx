import * as React from 'react';
import { ColumnWithProps } from 'components/Table';
import { CellProps } from 'react-table';
import { Link } from 'react-router-dom';
import { IAgency, IAgencyRecord } from 'interfaces';

export const columnDefinitions: ColumnWithProps<IAgencyRecord>[] = [
  {
    Header: 'Agency name',
    accessor: 'name',
    align: 'left',
    sortable: true,
    clickable: true,
  },
  {
    Header: 'Short name',
    accessor: 'code',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Description',
    accessor: 'description',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Parent Agency',
    accessor: 'parent',
    align: 'left',
    clickable: true,
    Cell: (props: CellProps<IAgency>) => {
      return (
        <Link to={`/admin/agency/${props.row.original.parentId}`}>{props.row.original.parent}</Link>
      );
    },
    sortable: true,
  },
];
