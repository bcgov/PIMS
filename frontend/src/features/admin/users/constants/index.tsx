import { ColumnWithProps } from 'components/Table';
import * as React from 'react';
import { FaCheck, FaTimes } from 'react-icons/fa';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IUserRecord } from '../interfaces/IUserRecord';

export const columnDefinitions: ColumnWithProps<IUserRecord>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IUserRecord>) => {
      return <Link to={`/admin/user/${props.row.original.id}`}>{props.row.original.username}</Link>;
    },
    sortable: true,
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Active',
    accessor: 'isDisabled',
    align: 'left',
    clickable: true,
    sortable: true,
    width: 100,
    Cell: (props: CellProps<IUserRecord>) =>
      props.row.original.isDisabled ? <FaTimes /> : <FaCheck />,
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    sortable: true,
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Roles',
    accessor: 'roles',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Last Login',
    accessor: 'lastLogin',
    align: 'left',
    clickable: true,
    minWidth: 100,
  },
  {
    Header: 'Created On',
    accessor: 'createdOn',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 100,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
