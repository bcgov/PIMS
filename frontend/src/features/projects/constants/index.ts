export * from './AgencyResponses';
export * from './DisposalWorkflows';
export * from './DisposeWorkflowStatus';
export * from './ProjectWorkflowComponents';
export * from './ReviewWorkflowStatus';
export * from './SPPApprovalTabs';
