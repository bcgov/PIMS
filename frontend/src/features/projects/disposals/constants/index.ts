export * from './defaultFilter';
export * from './defaultProjectForm';
