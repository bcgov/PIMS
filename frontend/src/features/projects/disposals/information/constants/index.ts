export * from './AddPropertyColumns';
export * from './PropertyColumns';
