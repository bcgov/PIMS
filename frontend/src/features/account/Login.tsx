import './Login.scss';
import React, { useState } from 'react';
import { Redirect } from 'react-router-dom';
import { Container, Row, Col, Button, Spinner, Jumbotron } from 'react-bootstrap';
import { useSelector } from 'react-redux';
import { IGenericNetworkAction } from 'actions/genericActions';
import { RootState } from 'reducers/rootReducer';
import { NEW_PIMS_USER } from 'actionCreators/usersActionCreator';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import * as actionTypes from 'constants/actionTypes';
import { useQuery } from 'hooks/use-query';
import { FaExternalLinkAlt } from 'react-icons/fa';
import PIMSlogo from 'assets/images/PIMSlogo/logo_with_text.png';

//check to see if user is using Internet Explorer
//as their browser
const usingIE = () => {
  const userAgent = window.navigator.userAgent;
  const isOldIE = userAgent.indexOf('MSIE '); //tag used for IE 10 or older
  const isIE11 = userAgent.indexOf('Trident/'); //tag used for IE11
  if (isOldIE > 0 || isIE11 > 0) return true;
  return false;
};

const Login = () => {
  const { redirect } = useQuery();
  const [showInstruction, setShowInstruction] = useState(false);
  const keyCloakWrapper = useKeycloakWrapper();
  const keycloak = keyCloakWrapper.obj;
  const isIE = usingIE();
  const activated = useSelector<RootState, IGenericNetworkAction>(
    state => (state.network as any)[actionTypes.ADD_ACTIVATE_USER] as IGenericNetworkAction,
  );
  if (!keycloak) {
    return <Spinner animation="border"></Spinner>;
  }
  if (keycloak?.authenticated) {
    if (activated?.status === NEW_PIMS_USER || !keyCloakWrapper?.roles?.length) {
      return <Redirect to={{ pathname: '/access/request' }} />;
    }
    return <Redirect to={redirect || '/mapview'} />;
  }
  if (isIE) {
    return <Redirect to={{ pathname: '/ienotsupported' }} />;
  }
  return (
    <Container className="login" fluid={true}>
      <Container className="unauth" fluid={true}>
        <img className="pims-logo" src={PIMSlogo} alt="PIMS logo" />
        <Row className="sign-in">
          <Col xs={1} md={3} />
          <Col xs={16} md={6} className="block">
            <h1>Search and visualize government property information</h1>
            <h6>
              PIMS enables you to search properties owned by the Government of British Columbia
            </h6>
            <p>
              The data provided can assist your agency in making informed, timely, and strategic
              decisions on the optimal use of real property assets on behalf of the people and
              priorities of British Columbia.
            </p>
            <Button variant="primary" onClick={() => keycloak.login()}>
              Sign In
            </Button>
            <p>Sign into PIMS with your government issued IDIR or with your Business BCeID.</p>
            <Row className="bceid">
              <Button variant="link" onClick={() => setShowInstruction(!showInstruction)}>
                Don't have a Business BCeID?
              </Button>
            </Row>
            <Row>
              {showInstruction && (
                <Jumbotron>
                  <p>
                    1. Search to see if your entity is{' '}
                    <a
                      href="https://www.bceid.ca/directories/whitepages"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      already registered
                    </a>{' '}
                    <FaExternalLinkAlt />
                  </p>
                  <p>
                    If you're not yet registered, <br></br>2.{' '}
                    <a
                      href="https://www.bceid.ca/os/?7169"
                      target="_blank"
                      rel="noopener noreferrer"
                    >
                      Register for your Business BCeID
                    </a>{' '}
                    <FaExternalLinkAlt />
                  </p>
                </Jumbotron>
              )}
            </Row>
          </Col>
          <Col xs={1} md={3} />
        </Row>
      </Container>
    </Container>
  );
};

export default Login;
