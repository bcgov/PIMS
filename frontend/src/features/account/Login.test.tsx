import React from 'react';
import { createMemoryHistory } from 'history';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { useKeycloak } from '@react-keycloak/web';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import Login from './Login';
import * as reducerTypes from 'constants/reducerTypes';
import {} from 'reducers/networkReducer';
import { IGenericNetworkAction } from 'actions/genericActions';
import { Provider } from 'react-redux';
import { ADD_ACTIVATE_USER } from 'constants/actionTypes';

jest.mock('axios');
jest.mock('@react-keycloak/web');
const mockStore = configureMockStore([thunk]);

const store = mockStore({
  [reducerTypes.NETWORK]: {
    [ADD_ACTIVATE_USER]: {},
  },
});

//boilerplate function used by most tests to wrap the Login component with a router.
const renderLogin = () => {
  const history = createMemoryHistory();
  return render(
    <Provider store={store}>
      <Router history={history}>
        <Login />
      </Router>
    </Provider>,
  );
};

describe('login', () => {
  afterEach(() => {
    cleanup();
  });
  it('login renders correctly', () => {
    (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { authenticated: false } });
    const history = createMemoryHistory();
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <Login></Login>
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('authenticated users are redirected to the mapview', () => {
    (useKeycloak as jest.Mock).mockReturnValue({
      keycloak: { authenticated: true, userInfo: { groups: ['System Administrator'] } },
    });
    const history = createMemoryHistory();

    render(
      <Provider store={store}>
        <Router history={history}>
          <Login />
        </Router>
      </Provider>,
    );
    expect(history.location.pathname).toBe('/mapview');
  });

  it('new users are sent to the guest page', () => {
    (useKeycloak as jest.Mock).mockReturnValue({
      keycloak: { authenticated: true, realmAccess: { roles: [{}] } },
    });
    const history = createMemoryHistory();
    const activatedAction: IGenericNetworkAction = {
      isFetching: false,
      name: 'test',
      type: 'POST',
      status: 201,
    };
    const store = mockStore({
      [reducerTypes.NETWORK]: {
        activateUser: activatedAction,
      },
    });

    render(
      <Provider store={store}>
        <Router history={history}>
          <Login />
        </Router>
      </Provider>,
    );
    expect(history.location.pathname).toBe('/access/request');
  });

  it('unAuthenticated users are shown the login screen', () => {
    (useKeycloak as jest.Mock).mockReturnValue({ keycloak: { authenticated: false } });
    const { getAllByRole } = renderLogin();
    expect(getAllByRole('heading')[0]).toHaveTextContent(
      'Search and visualize government property information',
    );
    expect(getAllByRole('heading')[1]).toHaveTextContent(
      'PIMS enables you to search properties owned by the Government of British Columbia',
    );
  });

  it('a spinner is displayed if keycloak has not yet been initialized', () => {
    (useKeycloak as jest.Mock).mockReturnValue({ keycloak: undefined });
    const { container } = renderLogin();
    expect(container.firstChild).toHaveClass('spinner-border');
  });

  it('the login button calls keycloaks login() method', () => {
    const login = jest.fn();
    (useKeycloak as jest.Mock).mockReturnValue({
      keycloak: { login: login, authenticated: false },
    });

    const { getByText } = renderLogin();
    fireEvent.click(getByText(/Sign In/));

    expect(login.mock.calls.length).toBe(1);
  });
});
