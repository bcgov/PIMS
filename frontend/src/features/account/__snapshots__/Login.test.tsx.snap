// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login login renders correctly 1`] = `
<div
  className="login container-fluid"
>
  <div
    className="unauth container-fluid"
  >
    <img
      alt="PIMS logo"
      className="pims-logo"
      src="logo_with_text.png"
    />
    <div
      className="sign-in row"
    >
      <div
        className="col-md-3 col-1"
      />
      <div
        className="block col-md-6 col-16"
      >
        <h1>
          Search and visualize government property information
        </h1>
        <h6>
          PIMS enables you to search properties owned by the Government of British Columbia
        </h6>
        <p>
          The data provided can assist your agency in making informed, timely, and strategic decisions on the optimal use of real property assets on behalf of the people and priorities of British Columbia.
        </p>
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Sign In
        </button>
        <p>
          Sign into PIMS with your government issued IDIR or with your Business BCeID.
        </p>
        <div
          className="bceid row"
        >
          <button
            className="btn btn-link"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Don't have a Business BCeID?
          </button>
        </div>
        <div
          className="row"
        />
      </div>
      <div
        className="col-md-3 col-1"
      />
    </div>
  </div>
</div>
`;
